feat(client): sign in/up messaging for migrating users

philbooth opened the pull request:
**Hopefully** fixes #2786, adding messages to the sign-in and sign-up views for migrating users.

Based on [this documentation](https://github.com/mozilla/fxa-content-server/blob/master/docs/query-params.md#migration), I predicated the new messages on the presence of a `migration` query parameter. I was unsure how to initiate a genuine migration scenario on my local machine though, so I've only tested it manually, by adding the query parameter to the URL myself. Hence I don't actually know if this change works end-to-end. How can I test that out?

It's also the first time I've touched the CSS, so I was really unsure about introducing the `info` class to `app/styles/_state.scss`. Have I missed some pre-existing class that I should use instead?

philbooth commented:
Circle CI failed, but I don't think it is something I've touched. This is the error:

```
× firefox on any platform - reset_password flow - verifying a reset password unlocks an account (32.558s)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/f49967ff-281d-4f53-85c3-788e6f7eb9d8/element / {"using":"css selector","value":"#fxa-settings-header"}] Unable to locate element: {"method":"css selector","selector":"#fxa-settings-header"}
For documentation on this error, please visit: http://seleniumhq.org/exceptions/no_such_element.html
Build info: version: '2.46.0', revision: '87c69e2', time: '2015-06-04 16:16:47'
System info: host: 'box738', ip: '127.0.0.1', os.name: 'Linux', os.arch: 'amd64', os.version: '3.14.28-031428-generic', java.version: '1.7.0_76'
Driver info: driver.version: unknown
  at ProxiedSession._post  <node_modules/intern/node_modules/leadfoot/Session.js:59:30>
  at ProxiedSession.Session.find  <node_modules/intern/node_modules/leadfoot/Session.js:1055:15>
  at Command.<anonymous>  <node_modules/intern/node_modules/leadfoot/Command.js:42:36>
  at <node_modules/intern/node_modules/dojo/Promise.ts:393:15>
  at Object.run [as _onImmediate]  <node_modules/intern/node_modules/dojo/Promise.ts:237:7>
  at processImmediate [as _immediateCallback]  <timers.js:345:15>
  at Command.find  <node_modules/intern/node_modules/leadfoot/Command.js:23:10>
  at Command.prototype.(anonymous function) [as findByCssSelector]  <node_modules/intern/node_modules/leadfoot/lib/strategies.js:24:16>
  at <tests/functional/reset_password.js:545:14>
  at onSettled  <app/bower_components/fxa-js-client/fxa-client.js:1312:9>
  at Object.flush [as _onImmediate]  <app/bower_components/fxa-js-client/fxa-client.js:1019:6>
  at processImmediate [as _immediateCallback]  <timers.js:345:15>
```

pdehaan commented on app/scripts/templates/sign_in.mustache:
```diff
@@ -19,6 +19,9 @@
       <div class="error"></div>
       <div class="success"></div>
+        {{#migration}}
+          <div class="info nudge">{{#t}}Already have a Firefox Account? Migrate your sync data by signing into it here.{{/t}}</div>
```
Nit: Not sure what "it" refers to here. FxA? Sync?
Also, should that be "you sync" or "your Sync"?

philbooth replied:
Fair points, I went verbatim from the mockup. @ryanfeeley?

shane-tomlinson replied:
@ryanfeeley - I know you always have good reasons, how come there is markedly different text between the signin and signup pages? Can the two pages have similar roots that diverge w/ the action taken?

signup: `Migrate your Sync data by creating a new Firefox Account`
signin: `Migrate your Sync data by signing in to an existing Firefox Account`

shane-tomlinson commented on app/scripts/views/sign_in.js:
```diff
@@ -63,7 +64,8 @@ function (Cocktail, p, BaseView, FormView, SignInTemplate, Session,
         password: this._formPrefill.get('password'),
-        error: this.error
+        error: this.error,
+        migration: this.isMigration()
```
Can you name `migration` `isMigration` (and in the templates too?)

shane-tomlinson commented on app/scripts/views/mixins/migration-mixin.js:
```diff
@@ -0,0 +1,18 @@
+
+  return {
+    isMigration: function () {
+      return !! Url.searchParam('migration', this.window.location.search);
```
We have slowly been moving away from the direct use of search parameters in the views because we found that importing data from the search params was inflexible when we wanted to use alternative data sources like the [ResumeToken](https://github.com/mozilla/fxa-content-server/blob/master/app/scripts/models/resume-token.js). Instead we have been using the AuthBroker and Relier models. All views have a reference to a `relier` model. We import `migration` into the [FxDesktopRelier](https://github.com/mozilla/fxa-content-server/blob/master/app/scripts/models/reliers/fx-desktop.js#L39). Similar to how we have [isFxDesktop](https://github.com/mozilla/fxa-content-server/blob/master/app/scripts/models/reliers/fx-desktop.js#L56), perhaps a new function `isMigration` could be used, or maybe line 15 could just read:

```js
   return this.relier.has('migration');
```

shane-tomlinson commented:
@ryanfeeley - we need yer help with copy. Copied from a now hidden comment:

----------------------------------

@ryanfeeley - I know you always have good reasons, how come there is markedly different text between the signin and signup pages? Can the two pages have similar roots that diverge w/ the action taken?

signup: `Migrate your Sync data by creating a new Firefox Account`
signin: `Migrate your Sync data by signing in to an existing Firefox Account`

ryanfeeley commented:
I tried to write the copy from the point of view of a user who already has a sync account and may be confusing that with a Firefox Account. We are getting higher failure rates on login, which leads me to believe that users are switching to login and trying their old sync credentials. I wanted to keep the copy in the positive (instead of i.e. "Your old sync account won't work"). Make sense and does it work?

philbooth commented:
@shane-tomlinson, I've pushed the relier change you requested.

One downside of that is I no longer know how to test/view the changes in my browser. Before I could just add the query string and see what it looked like, but I don't know how to do that now. Is there some special magic?

shane-tomlinson commented:
> Make sense and does it work?

@ryanfeeley - the reasoning makes sense, the signin text less so.

:bike: :shed: :rotating_light: (supposed to be bike-shed alert)

`Already have a Firefox Account? Migrate your sync data by signing into it here.`

`it` in `into it here` is vague. Is `it` Firefox Accounts, or is `it` Sync? 

`Migrate your sync data by signing into Firefox Accounts.` is to the point.

ryanfeeley commented:
Better, but I can out-bikeshed that:

`Migrate your sync data by signing in to your Firefox Account.`

It would be `in to` not `into` and though I don't like using the same word twice in a sentence I felt `your Firefox Account` was better than the plural `Firefox Accounts`.

philbooth commented:
Have re-pushed the sign up message as:

```
Migrate your sync data by signing in to your Firefox Account.
```

shane-tomlinson commented on app/tests/spec/views/sign_in.js:
```diff
@@ -146,6 +146,32 @@ function (chai, $, sinon, p, View, Session, AuthErrors, OAuthErrors, Metrics,
       });
+
+      it('displays a message if migration relier is set', function () {
+        sinon.stub(relier, 'has', function (arg) {
```
Since you have tested the `isMigration` functionality above, you could event just stub in `view.isMigration` here and ignore the relier.

shane-tomlinson commented:
The approach is solid! The `isMigration` stub comment above is a nit, not a blocker.

Two small comments:
* The blue box could use some vertical centering love.
* It'd be great to add functional tests in https://github.com/mozilla/fxa-content-server/blob/master/tests/functional/sync_sign_in.js and https://github.com/mozilla/fxa-content-server/blob/master/tests/functional/sync_sign_up.js. The tests could be nearly empty shells that just load up the pages with the `migration=sync1.1` query parameter and check to ensure the blue box is visible.

shane-tomlinson commented:
The vertical centering I was talking about:

![screen shot 2015-07-30 at 15 55 40](https://cloud.githubusercontent.com/assets/848085/8986543/90860dee-36d4-11e5-9112-6459ed62236b.png)

shane-tomlinson commented:
ref #2863. The strings added in #2863 should be removed with this PR.

philbooth commented:
> * The blue box could use some vertical centering love.

Yeah, I'd been trying to get by re-using existing styles as much as possible. I don't *think* that `nudge` class is actually in use any more though, certainly it doesn't appear elsewhere when I grep. So I've modified it directly and updated the comment to reflect its actual behaviour. Is that better?

> * It'd be great to add functional tests...

Done.
